### ISSUE TITLE
Dirichlet BC only works on a boundary with a submesh

### ISSUE THREAD

**Issue by baagaard-usgs** [geodynamics/pylith#220]
The Dirichlet BC assumes it can form a submesh for the boundary on which it is applied. In v2 we have different implementations for with and without a submesh. Can we add a switch to turn off the submesh stuff in v3?

The simplest test case is a mesh with one cell with the Dirichlet BC at a point (2D) or on an edge (3D).

**Comment by knepley** [comment]
At the lowest level, you specify a constraint on the Section, by saying how many dofs are affected and what the local dofs numbers are.

I usually go up one layer and use DMAddBoundary(). This just  takes a DMLabel, field, and components. If it is essential, it also takes a function used to generate boundary values.

**Comment by baagaard-usgs** [comment]
Yes, but we use an auxiliary field instead of a function. I think where things are breaking down now is the creation and discretization of the auxiliary field when it is not for a boundary. For such cases, do we just create a field over the domain for the auxiliary field instead?

**Comment by knepley** [comment]
You are allowed to give an FE field instead of a function, using DM_BOUNDARY_ESSENTIAL_FIELD.

**Comment by baagaard-usgs** [comment]
Right, but isn't the real problem what PETSc DM is associated with the auxiliary field (submesh or domain mesh)? If we don't have a boundary, we want the domain DM not a submesh DM, right?
